src/Microsoft.ML.Data/Evaluators/BinaryClassifierEvaluator.cs

```diff
@@ -828,6 +828,59 @@ public CalibratedBinaryClassificationMetrics Evaluate(IDataView data, string lab
             return result;
         }
 
+        /// <summary>
+        /// Evaluates scored binary classification data and generates precision recall curve data.
+        /// </summary>
+        /// <param name="data">The scored data.</param>
+        /// <param name="label">The name of the label column in <paramref name="data"/>.</param>
+        /// <param name="score">The name of the score column in <paramref name="data"/>.</param>
+        /// <param name="probability">The name of the probability column in <paramref name="data"/>, the calibrated version of <paramref name="score"/>.</param>
+        /// <param name="predictedLabel">The name of the predicted label column in <paramref name="data"/>.</param>
+        /// <param name="prCurve">The generated precision recall curve data. Up to 100000 of samples are used for p/r curve generation.</param>
+        /// <returns>The evaluation results for these calibrated outputs.</returns>
+        public CalibratedBinaryClassificationMetrics EvaluateWithPRCurve(IDataView data, string label, string score, string probability, string predictedLabel, out IEnumerable<PrecisionRecallMetrics> prCurve)
+        {
+            Host.CheckValue(data, nameof(data));
+            Host.CheckNonEmpty(label, nameof(label));
+            Host.CheckNonEmpty(score, nameof(score));
+            Host.CheckNonEmpty(probability, nameof(probability));
+            Host.CheckNonEmpty(predictedLabel, nameof(predictedLabel));
+
+            var roles = new RoleMappedData(data, opt: false,
+                RoleMappedSchema.ColumnRole.Label.Bind(label),
+                RoleMappedSchema.CreatePair(AnnotationUtils.Const.ScoreValueKind.Score, score),
+                RoleMappedSchema.CreatePair(AnnotationUtils.Const.ScoreValueKind.Probability, probability),
+                RoleMappedSchema.CreatePair(AnnotationUtils.Const.ScoreValueKind.PredictedLabel, predictedLabel));
+
+            var resultDict = ((IEvaluator)this).Evaluate(roles);
+            Host.Assert(resultDict.ContainsKey(MetricKinds.PrCurve));
+            var prCurveView = resultDict[MetricKinds.PrCurve];
+            Host.Assert(resultDict.ContainsKey(MetricKinds.OverallMetrics));
+            var overall = resultDict[MetricKinds.OverallMetrics];
+
+            List<PrecisionRecallMetrics> prCurveResult = new List<PrecisionRecallMetrics>();
+            using (var cursor = prCurveView.GetRowCursorForAllColumns())
+            {
+                while (cursor.MoveNext())
```

Comment:
Just to put in the PR what we had talked about, it is important that we get the getters *before* the tight loop, and we use the getters *inside* the tight loop to retrieve the values, since constructing the delegates is extremely expensive but using them is very cheap. (See e.g., [here](https://github.com/dotnet/machinelearning/blob/master/docs/code/IDataViewImplementation.md#why-getgetter).)

Reply:
thx

---
In reply to: [268291075](https://github.com/dotnet/machinelearning/pull/3039#discussion_r268291075) [](ancestors = 268291075)